File: Design-Patterns-Using-Swift/Iterator/IteratorPatternLinkedList.swift
class LinkedList: Sequence {
    // Node class representing elements of linked list.
    class Node {
        var value: Int
        var next: Node?

        init(_ value: Int, _ next: Node? = nil) {
            self.value = value
            self.next = next
        }
    }

    var head: Node?
    var tail: Node?
    var size: Int

    // Constructor of linked list.
    init() {
        head = nil
        tail = nil
        size = 0
    }

    func addHead(_ value: Int) {
        let newNode = Node(value, head)
        if head == nil {
            tail = newNode
        }
        head = newNode
        size += 1
    }

    func makeIterator() -> LinkedListIterator {
        return LinkedListIterator(aggregate: self)
    }

    func getSize() -> Int {
        return size
    }
}

struct LinkedListIterator: IteratorProtocol {
    var aggregate: LinkedList
    var current: LinkedList.Node?

    init(aggregate: LinkedList) {
        self.aggregate = aggregate
        self.current = aggregate.head
    }

    mutating func next() -> Int? {
        guard let current = current else {
            return nil
        }
        defer {
            self.current = current.next
        }
        return current.value
    }
}

// Client Code
var aggregate = LinkedList()
for i in 0..<5 {
    aggregate.addHead(i)
}

for val in aggregate {
    print(val, terminator: " ")
}

/*
4 3 2 1 0 
*/

File: Design-Patterns-Using-Cpp/Iterator/IteratorPatternLinkedList.cpp
#include <iostream>
#include <iterator>

class LinkedList {
private:
    class Node {
    public:
        int value;
        Node* next;

        Node(int v, Node* n) : value(v), next(n) {}
    };

    Node* head;
    Node* tail;
    int size;

public:
    LinkedList() : head(nullptr), tail(nullptr), size(0) {}

    void addHead(int value) {
        Node* newNode = new Node(value, head);
        if (head == nullptr) {
            tail = newNode;
        }
        head = newNode;
        size++;
    }

    class LinkedListIterator : public std::iterator<std::input_iterator_tag, int> {
    private:
        LinkedList* aggregate;
        Node* current;

    public:
        LinkedListIterator(LinkedList* aggregate, Node* current) : aggregate(aggregate), current(current) {}

        bool operator!=(const LinkedListIterator& other) const {
            return current != other.current && current != nullptr;
        }

        int operator*() const {
            return current->value;
        }

        LinkedListIterator& operator++() {
            if (current != nullptr) {
                current = current->next;
            }
            return *this;
        }
    };

    LinkedListIterator begin() {
        return LinkedListIterator(this, head);
    }

    LinkedListIterator end() {
        return LinkedListIterator(this, nullptr);
    }

    int getSize() const {
        return size;
    }
};

// Client code
int main() {
    LinkedList aggregate;
    for (int i = 0; i < 5; i++) {
        aggregate.addHead(i);
    }

    for (int val : aggregate) {
        std::cout << val << " ";
    }

    return 0;
}

/*
4 3 2 1 0 
*/